## Exploring Hackers News Posts

In this project, we'll compare two different types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Introduction
First, we'll read in the data and remove the headers.

In [31]:
# Read in the data.
opened_file = open('hacker_news.csv')

from csv import reader
hn = reader(opened_file)
hn = list(hn)
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing Headers

In [32]:
# Remove the headers
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN and Show HN Posts¶

In [34]:
# Create three empty lists called ask_posts, show_posts, and other_posts.
ask_posts = []
show_posts = []
other_posts = []

In [35]:
## Loop through each row in hn. Assign the title in each row to a variable named title
for row in hn:
    title = row[1]
    if (title.lower()).startswith('ask hn')==True:
        ask_posts.append(row)
    elif (title.lower()).startswith('show hn')==True:
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## Calculating the Average Number of Comments for Ask HN and Show HN Posts¶


In [36]:
# Find the total number of comments in ask posts 
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments = total_ask_comments + int(row[4])
print(total_ask_comments)

24483


In [37]:
# Compute the average number of comments on ask posts
avg_ask_comments = total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [38]:
# Compute total and avg comments on show posts
total_show_comments = 0
for row in show_posts:
    total_show_comments = total_show_comments + int(row[4])
print(total_show_comments)
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

11988
10.31669535283993


Finding: ask posts receive more comments on average.

## Finding the Amount of Ask Posts and Comments by Hour Created¶
Next, we'll determine if ask posts created at a certain time are more likely to attract comments.

In [39]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt
result_list = []
for row in ask_posts:
    result_list.append([row[6], int(row[4])])

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

## Calculating the Average Number of Comments for Ask HN Posts by Hour¶

In [42]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr]/counts_by_hour[hr]])
avg_by_hour

[['20', 21.525],
 ['22', 6.746478873239437],
 ['21', 16.009174311926607],
 ['19', 10.8],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['17', 11.46],
 ['10', 13.440677966101696],
 ['07', 7.852941176470588],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['03', 7.796296296296297],
 ['02', 23.810344827586206],
 ['14', 13.233644859813085],
 ['23', 7.985294117647059],
 ['16', 16.796296296296298],
 ['18', 13.20183486238532],
 ['15', 38.5948275862069],
 ['11', 11.051724137931034],
 ['01', 11.383333333333333],
 ['13', 14.741176470588234],
 ['04', 7.170212765957447],
 ['05', 10.08695652173913],
 ['12', 9.41095890410959]]

## Sorting and Printing Values from a List of Lists

In [41]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap

[[21.525, '20'], [6.746478873239437, '22'], [16.009174311926607, '21'], [10.8, '19'], [5.5777777777777775, '09'], [10.25, '08'], [11.46, '17'], [13.440677966101696, '10'], [7.852941176470588, '07'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.796296296296297, '03'], [23.810344827586206, '02'], [13.233644859813085, '14'], [7.985294117647059, '23'], [16.796296296296298, '16'], [13.20183486238532, '18'], [38.5948275862069, '15'], [11.051724137931034, '11'], [11.383333333333333, '01'], [14.741176470588234, '13'], [7.170212765957447, '04'], [10.08695652173913, '05'], [9.41095890410959, '12']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [29]:
# Sort the values and print the the 5 hours with the highest average comments.
print("Top 5 Hours for Ask Posts Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )
        

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.
According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

## Conclusion
Based on the above analysis, the recommendation for maximizing the number of comments and post receives is to create ask posts between 3 pm est - 4 pm est.